src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -0,0 +1,319 @@
+			Message<?> message = null;
+			try {
+				message = toMessagingMessage(record, acknowledgment, consumer);
+				if (KafkaInboundGateway.this.retryTemplate != null) {
+					message = addDeliveryAttemptHeader(message);
+				}
+				setAttributesIfNecessary(record, message);
+			}
+			catch (RuntimeException e) {
+				if (getErrorChannel() != null) {
+					KafkaInboundGateway.this.messagingTemplate.send(getErrorChannel(), buildErrorMessage(null,
+							new ConversionException("Failed to convert to message for: " + record, e)));
+				}
+			}
+			if (message != null) {
+				try {
+					Message<?> reply = sendAndReceiveMessage(message);
+					if (reply != null) {
+						reply = enhanceReply(message, reply);
+						KafkaInboundGateway.this.kafkaTemplate.send(reply);
+					}
+				}
+				finally {
+					if (KafkaInboundGateway.this.retryTemplate == null) {
+						attributesHolder.remove();
+					}
+				}
+			}
+			else {
+				KafkaInboundGateway.this.logger.debug("Converter returned a null message for: "
+						+ record);
+			}
+		}
+
+		private Message<?> addDeliveryAttemptHeader(Message<?> message) {
+			Message<?> messageToReturn = message;
+			AtomicInteger deliveryAttempt =
+					new AtomicInteger(((RetryContext) attributesHolder.get()).getRetryCount() + 1);
+			if (message.getHeaders() instanceof KafkaMessageHeaders) {
+				((KafkaMessageHeaders) message.getHeaders()).getRawHeaders()
+					.put(IntegrationMessageHeaderAccessor.DELIVERY_ATTEMPT, deliveryAttempt);
+			}
+			else {
+				messageToReturn = MessageBuilder.fromMessage(message)
+						.setHeader(IntegrationMessageHeaderAccessor.DELIVERY_ATTEMPT, deliveryAttempt)
+						.build();
+			}
+			return messageToReturn;
+		}
+
+		private Message<?> enhanceReply(Message<?> message, Message<?> reply) {
+			AbstractIntegrationMessageBuilder<?> builder = null;
+			MessageHeaders replyHeaders = reply.getHeaders();
+			MessageHeaders requestHeaders = message.getHeaders();
+			if (replyHeaders.get(KafkaHeaders.CORRELATION_ID) == null &&
+					requestHeaders.get(KafkaHeaders.CORRELATION_ID) != null) {
+				builder = getMessageBuilderFactory().fromMessage(reply)
+						.setHeader(KafkaHeaders.CORRELATION_ID,
+								message.getHeaders().get(KafkaHeaders.CORRELATION_ID));
+			}
+			if (replyHeaders.get(KafkaHeaders.TOPIC) == null &&
+					requestHeaders.get(KafkaHeaders.REPLY_TOPIC) != null) {
+				if (builder == null) {
+					builder = getMessageBuilderFactory().fromMessage(reply);
+				}
+				builder.setHeader(KafkaHeaders.TOPIC,
+						message.getHeaders().get(KafkaHeaders.REPLY_TOPIC));
+			}
+			if (replyHeaders.get(KafkaHeaders.PARTITION_ID) == null &&
+					requestHeaders.get(KafkaHeaders.REPLY_PARTITION) != null) {
+				if (builder == null) {
+					builder = getMessageBuilderFactory().fromMessage(reply);
+				}
+				builder.setHeader(KafkaHeaders.PARTITION_ID,
+						message.getHeaders().get(KafkaHeaders.REPLY_PARTITION));
+			}
+			if (builder != null) {
```

Comment:
Can't we just rely on the `MessageBuilder` API - using `setHeaderIfAbsent()`?

In the end it has a logic like `!this.modified` then return `originalMessage`.

Or do you worry about a void `MessageBuilder` instance as well?

Reply:
Yeah - no reason to create a builder unless we need one.